crates/bevy_render/src/shader/shader_reflect_naga.rs

```diff
@@ -0,0 +1,448 @@
+                            .iter_mut()
+                            .find(|buffer_desc: &&mut VertexBufferDescriptor| {
+                                buffer_desc.name.as_ref() == buffer_name
+                            }) {
+                            if current_buffer_desc_name.unwrap() != buffer_desc.name {
+                                panic!("vertex attribute buffer names must be consecutive")
+                            }
+
+                            buffer_desc
+                        } else {
+                            vertex_buffer_descriptors.push(VertexBufferDescriptor {
+                                name: buffer_name.to_owned().into(),
+                                stride: 0, // to be filled in later on
+                                step_mode,
+                                attributes: vec![],
+                            });
+                            vertex_buffer_descriptors.last_mut().unwrap()
+                        };
+
+                        current_buffer_desc_name = Some(buffer_desc.name.to_owned());
+
+                        buffer_desc.attributes.push(
+                            reflect_vertex_attribute_desc(&module, global, shader_location)
+                                .expect("unable to reflect vertex attributes"),
+                        );
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        // Sort the bind groups and attributes by set, binding, and location.
+        bind_groups.sort_unstable_by_key(|desc| desc.index);
+
+        for binding_desc in bind_groups.iter_mut().map(|desc| &mut desc.bindings[..]) {
+            binding_desc.sort_unstable_by_key(|desc| desc.index);
+        }
+
+        for buf_desc in vertex_buffer_descriptors.iter_mut() {
+            buf_desc
+                .attributes
+                .sort_unstable_by_key(|desc| desc.shader_location);
+
+            // Accumulate offsets and stride.
+            buf_desc.stride = buf_desc.attributes.iter_mut().fold(0, |offset, attr_desc| {
+                attr_desc.offset = offset;
+                offset + attr_desc.format.get_size()
+            });
+        }
+
+        ShaderLayout {
+            bind_groups,
+            vertex_buffer_descriptors,
+            entry_point: entry_point.name.clone(),
+        }
+    }
+}
+
+fn reflect_vertex_attribute_desc(
+    module: &naga::Module,
+    global: &naga::GlobalVariable,
+    shader_location: u32,
+) -> Result<VertexAttributeDescriptor, ()> {
+    use naga::{ScalarKind::*, TypeInner::*, VectorSize::*};
+
+    let ty = &module.types[global.ty];
+
+    let format = match ty.inner {
+        Scalar { kind, width } => match (kind, width) {
+            (Uint, 4) => VertexFormat::Uint,
+            (Sint, 4) => VertexFormat::Int,
+            (Float, 4) => VertexFormat::Float,
+            _ => return Err(()),
+        },
+        Vector { size, kind, width } => match (size, kind, width) {
```

Comment:
It would be great to have this (size, kind, width) information preserved. Take a look on my webgl2_renderer: https://github.com/bevyengine/bevy/pull/613/files#diff-d4ccfb870bfebbfae7f1b458a69aed63 - I have to recreate this information from this enum :)